trl/trainer/rloo_trainer.py

```diff
@@ -0,0 +1,723 @@
+        if args.deepspeed2:
+            reward_model = prepare_deepspeed2(reward_model, args.per_device_train_batch_size)
+            ref_policy = prepare_deepspeed2(ref_policy, args.per_device_train_batch_size)
+        elif args.deepspeed3:
+            reward_model = prepare_deepspeed3(reward_model, accelerator)
+            ref_policy = prepare_deepspeed3(ref_policy, accelerator)
+        else:
+            ref_policy = ref_policy.to(device)
+            reward_model = reward_model.to(device)
+
+        generation_config = GenerationConfig(
+            max_new_tokens=args.response_length,
+            min_new_tokens=args.response_length,
+            temperature=(args.temperature + 1e-7),
+            top_k=0.0,
+            top_p=1.0,
+            do_sample=True,
+        )
+
+        accelerator.print("===training policy===")
+        global_step = 0
+        start_time = time.time()
+        stats_shape = (args.num_ppo_epochs, args.num_mini_batches, args.gradient_accumulation_steps)
+        approxkl_stats = torch.zeros(stats_shape, device=device)
+        pg_clipfrac_stats = torch.zeros(stats_shape, device=device)
+        pg_loss_stats = torch.zeros(stats_shape, device=device)
+        vf_loss_stats = torch.zeros(stats_shape, device=device)
+        vf_clipfrac_stats = torch.zeros(stats_shape, device=device)
+        entropy_stats = torch.zeros(stats_shape, device=device)
+        ratio_stats = torch.zeros(stats_shape, device=device)
+        model.train()
+        for update in range(1, args.num_updates + 1):
+            global_step += 1 * args.batch_size
+            self.lr_scheduler.step()
+            data = next(iter_dataloader)
+            with torch.no_grad():
+                queries = data["input_ids"].to(device)
+                queries = queries.repeat(args.rloo_k, 1)
+                context_length = queries.shape[1]
+                query_responses = []
+                responses = []
+                postprocessed_responses = []
+                logprobs = []
+                ref_logprobs = []
+                scores = []
+                sequence_lengths = []
+                for i in range(0, queries.shape[0], args.local_rollout_forward_batch_size):
+                    query = queries[i : i + args.local_rollout_forward_batch_size]
+                    with unwrap_model_for_generation(model, self.accelerator) as unwrapped_model:
+                        query_response, logits = generate(
+                            unwrapped_model,
+                            query,
+                            tokenizer,
+                            generation_config,
+                        )
+                    response = query_response[:, context_length:]
+
+                    # use the logits during generation directly, instead of using the following
+                    all_logprob = F.log_softmax(logits, dim=-1)
+                    logprob = torch.gather(all_logprob, 2, response.unsqueeze(-1)).squeeze(-1)
```

Comment:
Why is `ref_logprob` scaled by `args.temperature + 1e-7`, but this `logprob` isn't?

Reply:
Great question. It's a limitation of the API, the logits from `generate` is already scaled.

To check it, feel free to print `logprob - ref_logprob`.